### PR TITLE
fix(connector): support kafka log compaction for debezium avro

### PR DESCRIPTION
**Pull request by KveinAxel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

* switch update event without before to insert
* delete with null payload

### <samp>🤖 Generated by Copilot at b9a3533</samp>

This pull request adds support for parsing Debezium events in Avro format, and handling tombstone and delete events as upsert operations. It modifies the `DebeziumAvro` parser in `avro_parser.rs` and the `is_upsert` method in `mod.rs` to enable this functionality. It also improves the comments and error handling of the parser.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [x] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Review by hzxa21 — COMMENTED** [review]
This PR solves part of the issue brought by log compaction by transforming UPDATE into INSERT but misses to solve the another parts:
- DELETE message can also have null before
- DELETE message's value itself can be null (the payload passed to `parse_inner` is empty) since we may only see the tombstone message after log compaction

**Comment by hzxa21** [comment]
See https://github.com/risingwavelabs/risingwave/issues/9267#issuecomment-1560447220 for my thoughts.

File: src/connector/src/parser/debezium/avro_parser.rs [hunk 1]
```diff
@@ -192,42 +193,42 @@ impl DebeziumAvroParser {
...
+            let (schema_id, mut raw_payload) = extract_schema_id(&key)?;
+            let key_schema = self.schema_resolver.get(schema_id).await?;
+            let key = from_avro_datum(key_schema.as_ref(), &mut raw_payload, None)
+                .map_err(|e| RwError::from(ProtocolError(e.to_string())))?;
+            return writer.delete(|column| {
+                let field_schema =
+                    extract_inner_field_schema(&self.inner_schema, Some(&column.name))?;
+                from_avro_value(
+                    get_field_from_avro_value(&key, column.name.as_str())?.clone(),
+                    field_schema,
```
Comment (tabVersion) [inline src/connector/src/parser/debezium/avro_parser.rs:211]: is it ok to fetch a non-exist column?

File: src/connector/src/source/kafka/source/reader.rs [hunk 2]
```diff
@@ -175,11 +175,7 @@ impl KafkaSplitReader {
                     Some(payload) => payload.len(),
                 };
                 num_messages += 1;
-                if self.enable_upsert {
-                    res.push(SourceMessage::from_kafka_message_upsert(msg));
-                } else {
-                    res.push(SourceMessage::from(msg));
-                }
+                res.push(SourceMessage::from(msg));
```
Comment (tabVersion) [inline src/connector/src/source/kafka/source/reader.rs:178]: lets keep the original ver, not all formats use upsert methods

**Review by tabVersion — APPROVED** [review]
rest LGTM

**Review by hzxa21 — APPROVED** [review]

File: scripts/source/test_data/debezium_compact_avro_json.1 [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+{"order_id": 1}^{"before": null, "after": {"Value": {"order_id": 1, "order_date": {"long": 1558430840000}, "customer_name": {"string": "Bob"}, "price": {"VariableScaleDecimal": {"scale": 1, "value": "i"}}, "product_id": {"int": 1}, "order_status": {"int": 1}}}, "source": {"version": "1.9.7.Final", "connector": "postgresql", "name": "postgres", "ts_ms": 1685975423896, "snapshot": {"string": "true"}, "db": "mydb", "sequence": {"string": "[null,\"23911624\"]"}, "schema": "public", "table": "orders", "txId": {"long": 490}, "lsn": {"long": 23911624}, "xmin": null}, "op": "r", "ts_ms": {"long": 1685975424096}, "transaction": null}
```
Comment (idx0-dev) [inline scripts/source/test_data/debezium_compact_avro_json.1:2]: It seems this part is not a valid json string.

**Review by idx0-dev — CHANGES_REQUESTED** [review]

**Pull request merged**